### DISCUSSION TITLE
Enabling time Module in Embedded MicroPython

### DISCUSSION THREAD

**Discussion by zdravko1989** [micropython/micropython#15876] [Other]
Hello everyone,

I am currently experimenting with the embedding example (see the examples/embedding directory). I'm trying to enable the time module, but I'm encountering some issues.

To enable the module, I modified the mpconfigport.h file with the following snippet:
```
#define MICROPY_PY_TIME                    (1)
#define MICROPY_PY_TIME_TIME_TIME_NS   (1)
#define MICROPY_PY_TIME_CUSTOM_SLEEP   (1)
#define MICROPY_PY_TIME_INCLUDEFILE    "ports/unix/modtime.c"
```
However, when I try to import the time module in my script, I receive the following error:

> File "<stdin>", in <module>
> ImportError: module not found

It seems that the time module is not being included in the build as expected.

Please note: the process in which I want to embed the MicroPython runtime will run on a Unix system.

Is there a specific step I’m missing in the configuration or build process to enable the time module in this setup? Any advice or guidance would be greatly appreciated!

Thanks in advance for your help!

**Comment by stinos** [comment]
Start with checking if the time code is effectively included in the build by introducing an error (just `#error` should do the trick) within extmod/modtime.c right before the last `#endif`

**Reply by zdravko1989** [reply]
No, it is not included. I still can build. Here is the mpconfigport.h that I am using

```
#include <port/mpconfigport_common.h>

#define MICROPY_CONFIG_ROM_LEVEL (MICROPY_CONFIG_ROM_LEVEL_MINIMUM)

#define MICROPY_ENABLE_COMPILER           (1)
#define MICROPY_ENABLE_GC                 (1)
#define MICROPY_FLOAT_IMPL                (MICROPY_FLOAT_IMPL_DOUBLE)
#define MICROPY_PY_SYS_PLATFORM           "linux"
#define MICROPY_PY_COLLECTIONS            (1)
#define MICROPY_PY_GC                     (1)
#define MICROPY_PY_GENERATOR_PEND_THROW   (1)
#define MICROPY_PY_MATH                   (1)
#define MICROPY_PY_STRUCT                 (1)
#define MICROPY_PY_SYS                    (1)
#define MICROPY_PY_SYS_PATH_ARGV_DEFAULTS (1)
#define MICROPY_PY_TIME                    (1)
#define MICROPY_PY_TIME_TIME_TIME_NS   (1)
#define MICROPY_PY_TIME_CUSTOM_SLEEP   (1)
#define MICROPY_PY_TIME_INCLUDEFILE    "extmod/modtime.c"
```

(I tried with "ports/unix/modtime.c" for  MICROPY_PY_TIME_INCLUDEFILE   as well)

And this is how I build embed port (I am including some custom modules as well):

```
  execute_process(
    COMMAND ${CMAKE_COMMAND} -E env
      make  MICROPYTHON_TOP=${MICROPYTHON_TOP} USER_C_MODULES=${USER_C_MODULES} 
      -f ${MICROPYTHON_TOP}/ports/embed/embed.mk
    )
```

**Reply by zdravko1989** [reply]
Looking in embed.mk (ports/embed) it seems that the embed package does include only modplatform.h from extmod and nothing else.
Right?

**Reply by stinos** [reply]
I just realize you could also just build with `make V=1` to get source files listed... 

Anyway: looks like embed.mk is super minimal, so you'll have to include the source file(s) then. Never used this but from the looks of it the idea is that all source files get copied into a target directory and then you include that one in the build somehow? So adding a `$(Q)$(CP) $(TOP)/extmod/modtime.c $(PACKAGE_DIR)/extmod` on line 56 in embed.mk or in your own makefile somewhere should do the trick?


> I tried with "ports/unix/modtime.c" for MICROPY_PY_TIME_INCLUDEFILE as wel

That is the only correct approach. this file gets included in extmod/modtime.c so `MICROPY_PY_TIME_INCLUDEFILE    "extmod/modtime.c"` would try to include itself..

**Reply by zdravko1989** [reply]
I tried this, but I ended up in having more and more dependencies. It seems that @cwalther found a way. 
Thanks a lot for your support!

**Comment by cwalther** [comment]
See #11430 for how I enabled the time module and others.

**Reply by zdravko1989** [reply]
Your full example works perfectly.
**Thank you so much!**

For now, I want to keep things minimalistic by adding only the `time`, `sys`, and a few custom C/C++ modules. I'll need to cherry-pick the necessary changes into my project, but that shouldn't be an issue.

Just out of curiosity, do you have a rough idea of when your PR might be merged?

**Reply by zdravko1989** [reply]
I don't have an issue with importing the time module, but I encountered a problem when trying this:

```
static const char *example_3 =
    "import time\n"
    "for i in range(10):  # Loop 10 times\n"
    "    print(f\"Execution {i + 1}\")\n"
    "    time.sleep(1)  # Pause for 1 second\n"
    "\n"
    "print('finish')\n"
    ;
```
When I ran the code, it executed much faster than expected, finishing in significantly less than 1 second. 
I had anticipated that the program would take at least 10 seconds to complete, given that the loop should pause for 1 second during each of its 10 iterations.

Do you have an idea what is wrong with the snippet?

**Reply by cwalther** [reply]
> do you have a rough idea of when your PR might be merged?

That depends on two things – when I get around to cleaning it up and taking it out of draft state, and when the maintainers get around to reviewing it. For both, my current answer is “no idea”. I haven’t been working on it for a while and don’t even remember offhand what the cleanups are that I still want to do.

> Do you have an idea what is wrong with the snippet?

Your snippet is fine, what’s missing is an implementation of the primitive behind `sleep()`, `mp_hal_delay_ms()` in [_mphal.c_](https://github.com/micropython/micropython/blob/d5673a51462f3bb06e758d54493dcaf8e0d6e855/examples/embedding-full/mphal.c#L42). What is there in my example is just a stub (like most of the rest of _mphal.c_) that does not actually sleep, because the actual implementation will depend on what environment you are embedding MicroPython into. For a Unix environment, you can probably take the implementation from _ports/unix/unix_mphal.c_.
